api/src/main/java/marquez/service/DatasetService.java

```diff
@@ -292,20 +324,67 @@ private Dataset toDataset(
             .map(TagName::of)
             .collect(toImmutableSet());
     final DatasetVersionRow versionRow =
-        versionDao
+        datasetVersionDao
             .find(
                 datasetRow.getType(),
                 (datasetVersionUuid == null)
                     ? datasetRow.getCurrentVersionUuid().orElse(null)
                     : datasetVersionUuid)
             .get();
     final ImmutableList<Field> fields =
-        fieldDao.findAllIn(toArray(versionRow.getFieldUuids(), UUID.class)).stream()
+        datasetFieldDao.findAllIn(toArray(versionRow.getFieldUuids(), UUID.class)).stream()
             .map(this::toField)
             .collect(toImmutableList());
     return Mapper.toDataset(datasetRow, tags, versionRow, fields);
   }
 
+  private DatasetVersion toDatasetVersion(@NonNull DatasetVersionRow datasetVersionRow) {
+    final ExtendedDatasetRow datasetRow =
+        datasetDao.findBy(datasetVersionRow.getDatasetUuid()).get();
+    final ImmutableSet<TagName> tags =
+        tagDao.findAllIn(toArray(datasetRow.getTagUuids(), UUID.class)).stream()
+            .map(TagRow::getName)
+            .map(TagName::of)
+            .collect(toImmutableSet());
+    final ImmutableList<Field> fields =
+        datasetFieldDao.findAllIn(toArray(datasetVersionRow.getFieldUuids(), UUID.class)).stream()
+            .map(this::toField)
+            .collect(toImmutableList());
+    final Run createdByRun =
+        datasetVersionRow
+            .getRunUuid()
+            .map(runUuid -> runService.getRun(RunId.of(runUuid)).orElse(null))
+            .orElse(null);
+    return Mapper.toDatasetVersion(datasetRow, tags, datasetVersionRow, fields, null, createdByRun);
+  }
+
+  private DatasetVersion toDatasetVersion(@NonNull ExtendedDatasetVersionRow datasetVersionRow) {
```

Comment:
This should be in the dao so this can be handled in one connection to put less contention of the connection pool, otherwise it will create n * ~4 connections per query. Since the limit is 100, it'd completely fill the connection pool if the full object is returned.

Reply:
Yeah, working with the current `dao` APIs, this was the "easiest" approach (definitely not ideal though). We'll want to start pushing more and more of this logic down to the database which is outlined in your [Object Model Rearchitecture Proposal ](https://github.com/MarquezProject/marquez/issues/912). We'll want to start moving towards a model outlined in your proposal.